includes/amp-canonical-actions.php

```diff
@@ -0,0 +1,108 @@
+		$content_max_width = $GLOBALS['content_width'];
+	}
+
+	$amp_content = new AMP_Content( $post->post_content,
+		array(
+			'AMP_Twitter_Embed_Handler' => array(),
+			'AMP_YouTube_Embed_Handler' => array(),
+			'AMP_Instagram_Embed_Handler' => array(),
+			'AMP_Vine_Embed_Handler' => array(),
+			'AMP_Facebook_Embed_Handler' => array(),
+			'AMP_Gallery_Embed_Handler' => array(),
+		),
+		array(
+			 'AMP_Blacklist_Sanitizer' => array(),
+			 'AMP_Img_Sanitizer' => array(),
+			 'AMP_Video_Sanitizer' => array(),
+			 'AMP_Audio_Sanitizer' => array(),
+			 'AMP_Iframe_Sanitizer' => array(
+				 'add_placeholder' => true,
+			 ),
+		),
+		array(
+			'content_max_width' => $content_max_width,
+		)
+	);
+
+	return $amp_content;
+
+}
+
+// Generate the AMP post content early on (runs the_content filters but skips our filter below)
+$GLOBALS['amp_content'] = amp_canonical_retrieve_content();
+
+// "the_content" filter was already invoked now, so attempt remove all filters
+remove_all_filters('the_content');
+
+// Callbacks for adding AMP-related things to the main theme when used as canonical
+add_action( 'wp_head', 'amp_canonical_add_scripts' );
+add_action( 'wp_head', 'amp_canonical_add_boilerplate_css' );
+add_action( 'wp_footer', 'amp_deregister_scripts' );
+
+// the final filter that replaces the content, run at the very end just to
+// make sure that no content filters run after it
+add_filter( 'the_content', 'amp_the_content_filter', PHP_INT_MAX);
+
+function amp_the_content_filter($content) {
+	if(isset($GLOBALS['amp_content'])) {
+		return $GLOBALS['amp_content']->get_amp_content();
+	} else {
+		return $content;
+	}
+}
+
+function amp_deregister_scripts() {
+  wp_deregister_script( 'wp-embed' );
+}
+
+function amp_canonical_add_boilerplate_css() {
+	?>
+	<style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
+	<?php
+}
+
+function amp_canonical_add_scripts() {
+
+	$amp_runtime_script = 'https://cdn.ampproject.org/v0.js';
+
+	// Always include AMP form & analytics, as almost every template has search form somewhere and is tracking page views
+	$scripts = array_merge(
+		array('amp-form' => 'https://cdn.ampproject.org/v0/amp-form-0.1.js'),
+		array('amp-analytics' => 'https://cdn.ampproject.org/v0/amp-analytics-0.1.js'),
```

Comment:
Should we leave this to themes/plugins to manage instead of assuming that all pages will have a form and/or analytics?